test/test_dirichlet_dist.cpp

```diff
@@ -0,0 +1,258 @@
+
+  alpha[0] = 1.36;
+  alpha[1] = 0.0; // alpha[1] = 0.
+  x[0] = 0.47;
+  x[1] = 0.53;
+  BOOST_MATH_CHECK_THROW(boost::math::pdf(dirichlet_distribution<V>(std::move(alpha)), x), std::domain_error);
+
+  alpha[0] = 1.26;
+  alpha[1] = 2.99;
+  x[0] = 0.5;
+  x[1] = 0.75; // sum(x) > 1.0
+  BOOST_MATH_CHECK_THROW(boost::math::pdf(dirichlet_distribution<V>(std::move(alpha)), x), std::domain_error);
+
+  alpha[0] = 1.56;
+  alpha[1] = 4.00;
+  x[0] = 0.31;
+  x[1] = -0.03; // x[1] < 0.
+  BOOST_MATH_CHECK_THROW(boost::math::pdf(dirichlet_distribution<V>(std::move(alpha)), x), std::domain_error);
+
+  alpha[0] = 1.56;
+  alpha[1] = 4.00;
+  x[0] = 0.31;
+  x[1] = 1.06; // x[1] > 1.
+  BOOST_MATH_CHECK_THROW(boost::math::pdf(dirichlet_distribution<V>(std::move(alpha)), x), std::domain_error);
+
+  // Domain test for cdf. Necessary conditions for cdf:
+  // 1. alpha[i] > 0
+  // 2. 0 <= x[i] <= 1
+  // 3. sum(x) <= 1.
+  alpha[0] = 1.56;
+  alpha[1] = 4.00;
+  x[0] = 0.31;
+  x[1] = 1.06; // x[1] > 1.
+  BOOST_MATH_CHECK_THROW(boost::math::cdf(dirichlet_distribution<V>(std::move(alpha)), x), std::domain_error);
+
+  alpha[0] = 3.756;
+  alpha[1] = 4.91;
+  x[0] = 0.31;
+  x[1] = -1.06; // x[1] < 0.
+  BOOST_MATH_CHECK_THROW(boost::math::cdf(dirichlet_distribution<V>(std::move(alpha)), x), std::domain_error);
+
+  alpha[0] = 1.56;
+  alpha[1] = -4.00; // alpha[1] < 0
+  x[0] = 0.31;
+  x[1] = 0.69;
+  BOOST_MATH_CHECK_THROW(boost::math::cdf(dirichlet_distribution<V>(std::move(alpha)), x), std::domain_error);
+
+  alpha[0] = 0.0;
+  alpha[1] = 4.00; // alpha[0] = 0.
+  x[0] = 0.25;
+  x[1] = 0.75;
+  BOOST_MATH_CHECK_THROW(boost::math::cdf(dirichlet_distribution<V>(std::move(alpha)), x), std::domain_error);
+
+  alpha[0] = 1.56;
+  alpha[1] = 4.00;
+  x[0] = 0.31;
+  x[1] = 0.71; // sum(x) > 1.
+  BOOST_MATH_CHECK_THROW(boost::math::cdf(dirichlet_distribution<V>(std::move(alpha)), x), std::domain_error);
+
+  // Domain test for mode. Necessary conditions for mode:
+  // 1. alpha[i] > 1.
+  alpha[0] = 1.0;
+  alpha[1] = 1.4; // alpha[0] = 1.
+  BOOST_MATH_CHECK_THROW(boost::math::mode(dirichlet_distribution<V>(std::move(alpha))), std::domain_error);
+
+  alpha[0] = 1.56;
+  alpha[1] = 0.92; // alpha[1] < 1.
+  BOOST_MATH_CHECK_THROW(boost::math::mode(dirichlet_distribution<V>(std::move(alpha))), std::domain_error);
+
+  // Some exact values of pdf.
+  alpha[0] = 1.0, alpha[1] = 1.0;
+  x[0] = 0.5, x[1] = 0.5;
+  BOOST_CHECK_EQUAL(boost::math::pdf(dirichlet_distribution<V>(std::move(alpha)), x), static_cast<RealType>(1.0));
+
+  alpha[0] = 2.0, alpha[1] = 2.0;
+  x[0] = 0.5, x[1] = 0.5;
+  BOOST_CHECK_EQUAL(boost::math::pdf(dirichlet_distribution<V>(std::move(alpha)), x), static_cast<RealType>(1.5));
+
+  // Checking precalculated values on scipy.
+  // https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.dirichlet.html
+  alpha[0] = static_cast<RealType>(5.778238829L); alpha[1] = static_cast<RealType>(2.55821892973L);
```

Comment:
Have these values been computed up to long double precision? If so it doesn't look like there's enough digits there. In fact, I don't think there are enough digits even for double precision.

Reply:
Scipy's gamma function doesn't support some (high precision) data types. Yes, these are not long double or double.

Reply:
Ok, I'd try to get some more digits here, maybe via Mathematica?

Reply:
I'd also try to find some exact values.